File: Factory_Pattern/fpm.cpp
#include <iostream>
#include <memory>
#include <vector>

class Pizza{
public:
	std::string name, dough, sauce;
	std::vector<std::string> toppings;

	void prepare(){
		std::cout<<"Preparing "<<name<<"\n";
		std::cout<<"Tossing dough...\n";
		std::cout<<"Adding sauce...\n";
		std::cout<<"Adding toppings: ";
		for(const std::string& s: toppings){
			std::cout<<" "<<s;
		}
	}
	virtual void bake(){
		std::cout<<"Bake for 25 min at 350\n";
	}
	virtual void cut(){
		std::cout<<"Cuttong the pizza into diagonal slices\n";
	}
	virtual void box(){
		std::cout<<"Place Pizza in official pizzaStore box\n";
	}

	std::string getName() {
		return this->name;
	}
};

class NYStyleCheesePizza: public Pizza {
public:
	explicit NYStyleCheesePizza() {
		this->name = "NY stule Sauce and Cheese Pizza";
		this->dough = "Thisn Crust Dough";
		this->sauce = "Marinara Sauce";
		this->toppings.push_back("grated Reggino Cheese");
	}
};

class NYStyleVeggiePizza: public Pizza {
public:
	explicit NYStyleVeggiePizza() {
		this->name = "Chicago style deep dish cheese pizza";
		this->dough = "Extra Thick Crust Dough";
		this->sauce = "Plum Tomoto sauce";
		this->toppings.push_back("Shredded Mozzarella Cheese");
	}

	void cut() override {
		std::cout<<"Cutting the pizza into square slices\n";
	}
};

class PizzaStore {
public:
	std::unique_ptr<Pizza> orderPizza(std::string pizza_type){
		std::unique_ptr<Pizza> pizza = std::make_unique<Pizza>();
		pizza = this->createPizza(pizza_type);
		pizza->prepare();
		pizza->bake();
		pizza->cut();
		pizza->box();
		return pizza;
	}

	virtual std::unique_ptr<Pizza> createPizza(std::string pizzaType) = 0; 
};


class NYPizzaStore : public PizzaStore {
	std::unique_ptr<Pizza> createPizza(std::string pizzaType) override {
		if (pizzaType == "Cheese"){
			return std::make_unique<NYStyleCheesePizza>();
		} else if (pizzaType == "Veggie") {
			return std::make_unique<NYStyleVeggiePizza>();
		} return nullptr;
	}
};

int main() {
	std::unique_ptr<PizzaStore> nyStore = std::make_unique<NYPizzaStore>();
	std::unique_ptr<Pizza> nyPizza = nyStore->orderPizza("Cheese");
	std::cout<<"Ethan ordered a "<<nyPizza->getName()<<"\n";
}

File: Factory_Pattern/fpm.py
from abc import ABC, abstractmethod

class Pizza(ABC):
    def __init__(self) -> None:
        super().__init__()
        self.name: str = None
        self.dough: str = None
        self.sauce: str = None
        self.toppings: list[str] = []

    def prepare(self) -> None:
        print(f'Prepare {self.name}')
        print('Tossing dough...')
        print('Adding sauce...')
        print(f'Adding topping: {self.toppings}')

    def bake(self) -> None: print('Bake for 25 min at 350')
    def cut(self) -> None: print('Cutting the pizza into diagonal slices')
    def box(self) -> None: print('Place pizza in official pizzastore box')
    def getName(self) -> None: return self.name


class NYStyleCheesePizza(Pizza):
    def __init__(self) -> None:
        super().__init__()
        self.name: str = "NY Style Sauce and Cheese Pizza"
        self.dough: str = "Thin Crust Dough"
        self.sauce: str = "Marinara Sauce"
        self.toppings.append("Grated Reggiano Cheese")

class ChicagoStyleCheesePizza(Pizza):
    def __init__(self) -> None:
        super().__init__()
        self.name: str = "Chicago Style Deep Dish Cheese Pizza"
        self.dough: str = "Extra Thick Crust Dough"
        self.sauce: str = "Plum Tomato Sauce"
        self.toppings.append("Shredded Mozzarella Cheese")

    def cut(self) -> None:
        print("Cutting the pizza into square slices")

class PizzaStore(ABC):
    def orderPizza(self, pizzaType: str) -> Pizza:
        pizza: Pizza =  self.createPizza(pizzaType)
        pizza.prepare()
        pizza.bake()
        pizza.cut()
        pizza.box()
        return pizza

    @abstractmethod
    def createPizza(self, pizzaType: str) -> Pizza: pass


class NYPizzaStore(PizzaStore):
    def createPizza(self, pizzaType: str) -> Pizza:
        if pizzaType == "Cheese":
            return NYStyleCheesePizza()
        else:
            return None


def main() -> None:
    nyStore = NYPizzaStore()
    pizza = nyStore.orderPizza("Cheese")
    print(f"Ethan ordered a {pizza.getName()}")


if __name__ == '__main__':
    main()
    
        